File: server-java-play-framework/app/controllers/BidApiController.java
package controllers;

import apimodels.BidResponse;
import java.math.BigDecimal;
import apimodels.SirqulResponse;

import com.typesafe.config.Config;
import play.mvc.Controller;
import play.mvc.Result;
import play.mvc.Http;
import java.util.List;
import java.util.Map;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.JsonNode;
import com.google.inject.Inject;
import java.io.File;
import play.libs.Files.TemporaryFile;
import openapitools.OpenAPIUtils;
import com.fasterxml.jackson.core.type.TypeReference;

import javax.validation.constraints.*;
import javax.validation.Valid;
import com.typesafe.config.Config;

import openapitools.OpenAPIUtils.ApiAction;

@javax.annotation.Generated(value = "org.openapitools.codegen.languages.JavaPlayFrameworkCodegen", date = "2026-02-09T20:49:20.380387445Z[Etc/UTC]", comments = "Generator version: 7.20.0-SNAPSHOT")
public class BidApiController extends Controller {
    private final BidApiControllerImpInterface imp;
    private final ObjectMapper mapper;
    private final Config configuration;

    @Inject
    private BidApiController(Config configuration, BidApiControllerImpInterface imp) {
        this.imp = imp;
        mapper = new ObjectMapper();
        this.configuration = configuration;
    }

    @ApiAction
    public Result createBid(Http.Request request, BigDecimal version) throws Exception {
        String valuedeviceId = request.getQueryString("deviceId");
        String deviceId;
        if (valuedeviceId != null) {
            deviceId = valuedeviceId;
        } else {
            deviceId = null;
        }
        String valueaccountId = request.getQueryString("accountId");
        Long accountId;
        if (valueaccountId != null) {
            accountId = Long.parseLong(valueaccountId);
        } else {
            accountId = null;
        }
        String valuebiddableType = request.getQueryString("biddableType");
        String biddableType;
        if (valuebiddableType != null) {
            biddableType = valuebiddableType;
        } else {
            throw new IllegalArgumentException("'biddableType' parameter is required");
        }
        String valuebiddableId = request.getQueryString("biddableId");
        Long biddableId;
        if (valuebiddableId != null) {
            biddableId = Long.parseLong(valuebiddableId);
        } else {
            throw new IllegalArgumentException("'biddableId' parameter is required");
        }
        String valueamountPerView = request.getQueryString("amountPerView");
        Double amountPerView;
        if (valueamountPerView != null) {
            amountPerView = Double.parseDouble(valueamountPerView);
        } else {
            throw new IllegalArgumentException("'amountPerView' parameter is required");
        }
        String valueamountPerAction = request.getQueryString("amountPerAction");
        Double amountPerAction;
        if (valueamountPerAction != null) {
            amountPerAction = Double.parseDouble(valueamountPerAction);
        } else {
            throw new IllegalArgumentException("'amountPerAction' parameter is required");
        }
        String valuebudgetAmount = request.getQueryString("budgetAmount");
        Double budgetAmount;
        if (valuebudgetAmount != null) {
            budgetAmount = Double.parseDouble(valuebudgetAmount);
        } else {
            throw new IllegalArgumentException("'budgetAmount' parameter is required");
        }
        String valuebudgetSchedule = request.getQueryString("budgetSchedule");
        String budgetSchedule;
        if (valuebudgetSchedule != null) {
            budgetSchedule = valuebudgetSchedule;
        } else {
            throw new IllegalArgumentException("'budgetSchedule' parameter is required");
        }
        return imp.createBidHttp(request, version, biddableType, biddableId, amountPerView, amountPerAction, budgetAmount, budgetSchedule, deviceId, accountId);
    }

    @ApiAction
    public Result deleteBid(Http.Request request, BigDecimal version) throws Exception {
        String valuedeviceId = request.getQueryString("deviceId");
        String deviceId;
        if (valuedeviceId != null) {
            deviceId = valuedeviceId;
        } else {
            deviceId = null;
        }
        String valueaccountId = request.getQueryString("accountId");
        Long accountId;
        if (valueaccountId != null) {
            accountId = Long.parseLong(valueaccountId);
        } else {
            accountId = null;
        }
        String valuebidId = request.getQueryString("bidId");
        Long bidId;
        if (valuebidId != null) {
            bidId = Long.parseLong(valuebidId);
        } else {
            throw new IllegalArgumentException("'bidId' parameter is required");
        }
        return imp.deleteBidHttp(request, version, bidId, deviceId, accountId);
    }

    @ApiAction
    public Result getBid(Http.Request request, BigDecimal version) throws Exception {
        String valuedeviceId = request.getQueryString("deviceId");
        String deviceId;
        if (valuedeviceId != null) {
            deviceId = valuedeviceId;
        } else {
            deviceId = null;
        }
        String valueaccountId = request.getQueryString("accountId");
        Long accountId;
        if (valueaccountId != null) {
            accountId = Long.parseLong(valueaccountId);
        } else {
            accountId = null;
        }
        String valuebidId = request.getQueryString("bidId");
        Long bidId;
        if (valuebidId != null) {
            bidId = Long.parseLong(valuebidId);
        } else {
            throw new IllegalArgumentException("'bidId' parameter is required");
        }
        return imp.getBidHttp(request, version, bidId, deviceId, accountId);
    }

    @ApiAction
    public Result updateBid(Http.Request request, BigDecimal version) throws Exception {
        String valuedeviceId = request.getQueryString("deviceId");
        String deviceId;
        if (valuedeviceId != null) {
            deviceId = valuedeviceId;
        } else {
            deviceId = null;
        }
        String valueaccountId = request.getQueryString("accountId");
        Long accountId;
        if (valueaccountId != null) {
            accountId = Long.parseLong(valueaccountId);
        } else {
            accountId = null;
        }
        String valuebidId = request.getQueryString("bidId");
        Long bidId;
        if (valuebidId != null) {
            bidId = Long.parseLong(valuebidId);
        } else {
            throw new IllegalArgumentException("'bidId' parameter is required");
        }
        String valueamountPerView = request.getQueryString("amountPerView");
        Double amountPerView;
        if (valueamountPerView != null) {
            amountPerView = Double.parseDouble(valueamountPerView);
        } else {
            amountPerView = null;
        }
        String valueamountPerAction = request.getQueryString("amountPerAction");
        Double amountPerAction;
        if (valueamountPerAction != null) {
            amountPerAction = Double.parseDouble(valueamountPerAction);
        } else {
            amountPerAction = null;
        }
        String valuebudgetAmount = request.getQueryString("budgetAmount");
        Double budgetAmount;
        if (valuebudgetAmount != null) {
            budgetAmount = Double.parseDouble(valuebudgetAmount);
        } else {
            budgetAmount = null;
        }
        String valuebudgetSchedule = request.getQueryString("budgetSchedule");
        String budgetSchedule;
        if (valuebudgetSchedule != null) {
            budgetSchedule = valuebudgetSchedule;
        } else {
            budgetSchedule = null;
        }
        return imp.updateBidHttp(request, version, bidId, deviceId, accountId, amountPerView, amountPerAction, budgetAmount, budgetSchedule);
    }

}

File: server-scala-play-server/app/api/BidApiController.scala
package api

import org.openapitools.OpenApiExceptions
import javax.inject.{Inject, Singleton}
import play.api.libs.json._
import play.api.mvc._
import model.BidResponse
import model.SirqulResponse

@javax.annotation.Generated(value = Array("org.openapitools.codegen.languages.ScalaPlayFrameworkServerCodegen"), date = "2026-02-11T19:31:43.310890579Z[Etc/UTC]", comments = "Generator version: 7.20.0-SNAPSHOT")
@Singleton
class BidApiController @Inject()(cc: ControllerComponents, api: BidApi) extends AbstractController(cc) {
  /**
    * POST /api/3.18/bid/create?deviceId=[value]&accountId=[value]&biddableType=[value]&biddableId=[value]&amountPerView=[value]&amountPerAction=[value]&budgetAmount=[value]&budgetSchedule=[value]
    */
  def createBid(): Action[AnyContent] = Action { request =>
    def executeApi(): BidResponse = {
      val deviceId = request.getQueryString("deviceId")
        
      val accountId = request.getQueryString("accountId")
        .map(value => value.toLong)
        
      val biddableType = request.getQueryString("biddableType")
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("biddableType", "query string")
        }
        
      val biddableId = request.getQueryString("biddableId")
        .map(value => value.toLong)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("biddableId", "query string")
        }
        
      val amountPerView = request.getQueryString("amountPerView")
        .map(value => value.toDouble)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("amountPerView", "query string")
        }
        
      val amountPerAction = request.getQueryString("amountPerAction")
        .map(value => value.toDouble)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("amountPerAction", "query string")
        }
        
      val budgetAmount = request.getQueryString("budgetAmount")
        .map(value => value.toDouble)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("budgetAmount", "query string")
        }
        
      val budgetSchedule = request.getQueryString("budgetSchedule")
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("budgetSchedule", "query string")
        }
        
      api.createBid(biddableType, biddableId, amountPerView, amountPerAction, budgetAmount, budgetSchedule, deviceId, accountId)
    }

    val result = executeApi()
    val json = Json.toJson(result)
    Ok(json)
  }

  /**
    * POST /api/3.18/bid/delete?deviceId=[value]&accountId=[value]&bidId=[value]
    */
  def deleteBid(): Action[AnyContent] = Action { request =>
    def executeApi(): SirqulResponse = {
      val deviceId = request.getQueryString("deviceId")
        
      val accountId = request.getQueryString("accountId")
        .map(value => value.toLong)
        
      val bidId = request.getQueryString("bidId")
        .map(value => value.toLong)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("bidId", "query string")
        }
        
      api.deleteBid(bidId, deviceId, accountId)
    }

    val result = executeApi()
    val json = Json.toJson(result)
    Ok(json)
  }

  /**
    * GET /api/3.18/bid/get?deviceId=[value]&accountId=[value]&bidId=[value]
    */
  def getBid(): Action[AnyContent] = Action { request =>
    def executeApi(): BidResponse = {
      val deviceId = request.getQueryString("deviceId")
        
      val accountId = request.getQueryString("accountId")
        .map(value => value.toLong)
        
      val bidId = request.getQueryString("bidId")
        .map(value => value.toLong)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("bidId", "query string")
        }
        
      api.getBid(bidId, deviceId, accountId)
    }

    val result = executeApi()
    val json = Json.toJson(result)
    Ok(json)
  }

  /**
    * POST /api/3.18/bid/update?deviceId=[value]&accountId=[value]&bidId=[value]&amountPerView=[value]&amountPerAction=[value]&budgetAmount=[value]&budgetSchedule=[value]
    */
  def updateBid(): Action[AnyContent] = Action { request =>
    def executeApi(): BidResponse = {
      val deviceId = request.getQueryString("deviceId")
        
      val accountId = request.getQueryString("accountId")
        .map(value => value.toLong)
        
      val bidId = request.getQueryString("bidId")
        .map(value => value.toLong)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("bidId", "query string")
        }
        
      val amountPerView = request.getQueryString("amountPerView")
        .map(value => value.toDouble)
        
      val amountPerAction = request.getQueryString("amountPerAction")
        .map(value => value.toDouble)
        
      val budgetAmount = request.getQueryString("budgetAmount")
        .map(value => value.toDouble)
        
      val budgetSchedule = request.getQueryString("budgetSchedule")
        
      api.updateBid(bidId, deviceId, accountId, amountPerView, amountPerAction, budgetAmount, budgetSchedule)
    }

    val result = executeApi()
    val json = Json.toJson(result)
    Ok(json)
  }

  private def splitCollectionParam(paramValues: String, collectionFormat: String): List[String] = {
    val splitBy =
      collectionFormat match {
        case "csv" => ",+"
        case "tsv" => "\t+"
        case "ssv" => " +"
        case "pipes" => "|+"
      }

    paramValues.split(splitBy).toList
  }
}
